src/triage/component/postmodeling/contrast/group_evaluator.py

```diff
@@ -0,0 +1,325 @@
+            FROM model_metadata.models m
+            JOIN model_metadata.model_groups mg
+            USING (model_group_id)
+            WHERE model_group_id IN {}
+        ),
+        individual_model_id_matrices AS(
+        SELECT DISTINCT ON (matrix_uuid)
+               model_id,
+               matrix_uuid
+            FROM test_results.predictions
+            WHERE model_id = ANY(
+                SELECT model_id
+                FROM individual_model_ids_metadata
+            )
+        )
+        SELECT metadata.model_id,
+               metadata.model_group_id,
+               metadata.model_hash,
+               metadata.hyperparameters,
+               metadata.train_end_time,
+               metadata.train_matrix_uuid,
+               metadata.training_label_timespan,
+               metadata.model_type,
+               metadata.model_config,
+               test.matrix_uuid AS test_matrix_uuid
+        FROM individual_model_ids_metadata AS metadata
+        LEFT JOIN individual_model_id_matrices AS test
+        USING(model_id);
+        """.format(self.model_group_id),
+                   con=db_engine).to_dict('list')
+
+        # Add metadata attributes to model
+        self.model_id = model_metadata['model_id']
+        self.model_type = model_metadata['model_type']
+        self.train_end_time = model_metadata['train_end_time']
+        self.hyperparameters = model_metadata['hyperparameters']
+        self.model_hash = model_metadata['model_hash']
+        self.train_matrix_uuid = model_metadata['train_matrix_uuid']
+        self.test_matrix_uuid = model_metadata['test_matrix_uuid']
+        self.train_matrix = None
+        self.pred_matrix = None
+        self.preds = None
+        self.feature_importances = None
+        self.model_metrics = None
+
+    def __str__(self):
+        s = 'Model collection object for model_ids: {}'.format(self.model_id)
+        s += '\n Model Groups: {}'.format(self.model_group_id)
+        s += '\n Model types: {}'.format(self.model_type)
+        s += '\n Model hyperparameters: {}'.format(self.hyperparameters)
+        s += '\Matrix hashes (train,test): {}'.format((self.train_matrix_uuid,
+                                                      self.test_matrix_uuid))
+        return s
+
+    def _load_predictions(self):
+        preds = pd.read_sql("""
+                SELECT model_id,
+                       entity_id,
+                       as_of_date,
+                       score,
+                       label_value,
+                       COALESCE(rank_abs, RANK() 
+                       OVER(PARTITION BY model_id ORDER BY score DESC)) AS rank_abs,
+                       rank_pct,
+                       test_label_timespan
+                FROM test_results.predictions
+                WHERE model_id IN {}
+                AND label_value IS NOT NULL
+                """.format(tuple(self.model_id)),
+                con=db_engine)
+        self.preds = preds
```

Comment:
just use `descriptors.cachedproperty`:

    @cachedproperty
    def preds(self):
        return pd.read_sql("…")

…and forget about this `_load_…` stuff :wink:

Reply:
Nice, I didn't know this trick

Reply:
It helps to write the library 😉